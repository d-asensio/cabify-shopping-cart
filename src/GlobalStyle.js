import normalizeStyle from 'styled-normalize'
import { createGlobalStyle, css } from 'styled-components'

const customStyleReset = css`
  ${normalizeStyle}

  * {
    box-sizing: border-box;

    &::before,
    &::after {
      box-sizing: border-box;
    }
  }

  button {
    font-family: Arial, "Helvetica Neue", Helvetica, sans-serif;
  }

  :root {
    font-family: Avenir, "Avenir Next", "Segoe UI", sans-serif;
    font-size: 16px;

    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
  }
`

const GlobalStyle = createGlobalStyle`
  ${customStyleReset}

  body {
    display: flex;
    align-items: center;
    justify-content: center;

    width: 100vw;
    height: 100vh;

    background-image: url("../images/background.png");
    background-color: #212240;
    background-position: center;
    background-size: cover;
  }
`

export default GlobalStyle
