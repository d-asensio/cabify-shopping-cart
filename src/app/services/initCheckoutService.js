import memoize from 'lodash.memoize'

import Checkout from '../../checkout'

const initCheckoutService = function ({ fetch = window.fetch } = {}) {
  const SERVICE_URL = '/data'

  let checkout = null

  const _getServiceData = memoize(async (dataPath) => {
    const response = await fetch(`${SERVICE_URL}/${dataPath}.json`)
    return response.json()
  })

  async function getAvailableProducts () {
    return _getServiceData('products')
  }

  async function getSummaryForSelection (selectedProducts) {
    await _initNewCheckout()

    _applySelectionToCheckout(selectedProducts)

    return {
      discounts: checkout.discountsBreakdown(),
      total: checkout.total()
    }
  }

  async function _initNewCheckout () {
    const products = await _getServiceData('products')
    const discounts = await _getServiceData('discounts')

    checkout = new Checkout({
      products,
      discounts
    })
  }

  function _applySelectionToCheckout (selectedProducts) {
    for (const selectedProduct of selectedProducts) {
      checkout.scan(
        selectedProduct.id,
        selectedProduct.quantity
      )
    }
  }

  return {
    getAvailableProducts,
    getSummaryForSelection
  }
}

export default initCheckoutService
