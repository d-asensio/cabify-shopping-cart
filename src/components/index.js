export { default as OrderSummary } from './OrderSummary'
export { default as ProductDetails } from './ProductDetails'
export { default as ProductList } from './ProductList'
export { default as SectionHeading } from './SectionHeading'
export { default as ShoppingCart } from './ShoppingCart'
export { default as Stepper } from './Stepper'
export { default as TableGrid } from './TableGrid'
export { default as Thumbnail } from './Thumbnail'
