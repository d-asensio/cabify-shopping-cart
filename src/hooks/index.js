export { default as useControlledInputNumber } from './useControlledInputNumber'
