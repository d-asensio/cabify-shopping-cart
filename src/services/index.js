export { default as initCheckoutService } from './initCheckoutService'
