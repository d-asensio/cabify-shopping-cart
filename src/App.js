import React from 'react'
import styled from 'styled-components'

import { ProductDetails } from './components'

const Wrapper = styled.main`
  display: flex;
  flex-direction: column;

  background-color: #fff;

  border-radius: 4px;
`

function App () {
  return (
    <Wrapper>
      <ProductDetails
        imageSrc='images/thumbnails/shirt.png'
        name='Shirt'
        code='X7R2OPX'
      />
      <ProductDetails
        imageSrc='images/thumbnails/mug.png'
        name='Mug'
        code='X7R2OPZ'
      />
      <ProductDetails
        imageSrc='images/thumbnails/cap.png'
        name='Cap'
        code='X7R2OPY'
      />
    </Wrapper>
  )
}

export default App
