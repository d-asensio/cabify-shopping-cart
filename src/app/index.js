import React from 'react'

import App from './App'
import GlobalStyle from './GlobalStyle'

import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
import createSagaMiddleware from 'redux-saga'

import { reducer } from './reducers'
import { saga } from './sagas'

import { initCheckoutService } from './services'

const checkoutService = initCheckoutService()

const sagaMiddleware = createSagaMiddleware({
  context: {
    checkoutService
  }
})

const store = configureStore({
  reducer,
  middleware: [
    sagaMiddleware
  ],
  devTools: true
})

sagaMiddleware.run(saga)

export default () => (
  <Provider store={store}>
    <GlobalStyle />
    <App />
  </Provider>
)
