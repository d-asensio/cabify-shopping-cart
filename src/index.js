import React from 'react'
import ReactDOM from 'react-dom'

import App from './App'
import GlobalStyle from './GlobalStyle'

import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
import createSagaMiddleware from 'redux-saga'

import { reducer } from './reducers'
import { saga } from './sagas'

const sagaMiddleware = createSagaMiddleware()

const store = configureStore({
  reducer,
  middleware: [
    sagaMiddleware
  ],
  devTools: true
})

sagaMiddleware.run(saga)

ReactDOM.render(
  <Provider store={store}>
    <GlobalStyle />
    <App />
  </Provider>,
  document.getElementById('app')
)
