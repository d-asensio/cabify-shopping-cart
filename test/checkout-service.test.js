import { Checkout } from '../src/services'

describe('Checkout service', () => {
  it('do not fail at initialization', () => {
    const checkout = new Checkout({
      products: [
        {
          id: 'TSHIRT',
          name: 'Shirt',
          code: 'X7R2OPX',
          price: 20.00
        }
      ]
    })

    expect(checkout).toBeInstanceOf(Checkout)
  })

  it('do not fail at scanning a valid product', () => {
    const checkout = new Checkout({
      products: [
        {
          id: 'TSHIRT',
          name: 'Shirt',
          code: 'X7R2OPX',
          price: 20.00
        }
      ]
    })

    expect(() => {
      checkout.scan('TSHIRT')
    }).not.toThrowError()
  })

  it('returns an instance of `Checkout` to enable chaining', () => {
    const checkout = new Checkout({
      products: [
        {
          id: 'TSHIRT',
          name: 'Shirt',
          code: 'X7R2OPX',
          price: 20.00
        }
      ]
    })

    expect(
      checkout.scan('TSHIRT')
    ).toBeInstanceOf(Checkout)
  })

  it('trows an exception at scanning a unexisting product', () => {
    const checkout = new Checkout({
      products: [
        {
          id: 'TSHIRT',
          name: 'Shirt',
          code: 'X7R2OPX',
          price: 20.00
        }
      ]
    })

    expect(() => {
      checkout.scan('MUG')
    }).toThrowErrorMatchingSnapshot()
  })

  it('returns a total amount of 0 if no products have been scanned', () => {
    const checkout = new Checkout({
      products: [
        {
          id: 'TSHIRT',
          name: 'Shirt',
          code: 'X7R2OPX',
          price: 20.00
        }
      ]
    })

    expect(
      checkout.total()
    ).toBe(0)
  })

  it('returns the correct total amount if one product have been scanned', () => {
    const checkout = new Checkout({
      products: [
        {
          id: 'TSHIRT',
          name: 'Shirt',
          code: 'X7R2OPX',
          price: 20.00
        }
      ]
    })

    expect(
      checkout
        .scan('TSHIRT')
        .total()
    ).toBe(20)
  })

  it('returns the correct total amount if one product have been scanned multiple times', () => {
    const checkout = new Checkout({
      products: [
        {
          id: 'TSHIRT',
          name: 'Shirt',
          code: 'X7R2OPX',
          price: 20.00
        }
      ]
    })

    expect(
      checkout
        .scan('TSHIRT')
        .scan('TSHIRT')
        .total()
    ).toBe(40)
  })

  it('returns the correct total amount if multiple different products have been scanned', () => {
    const checkout = new Checkout({
      products: [
        {
          id: 'TSHIRT',
          name: 'Shirt',
          code: 'X7R2OPX',
          price: 20.00
        },
        {
          id: 'MUG',
          name: 'Mug',
          code: 'X7R2OPY',
          price: 5.00
        },
        {
          id: 'CAP',
          name: 'Cap',
          code: 'X7R2OPZ',
          price: 10.00
        }
      ]
    })

    expect(
      checkout
        .scan('TSHIRT')
        .scan('MUG')
        .scan('CAP')
        .total()
    ).toBe(35)
  })
})
