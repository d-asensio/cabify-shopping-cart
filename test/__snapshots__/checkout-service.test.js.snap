// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkout service trows an exception at scanning a unexisting product 1`] = `"'The product identified by \\"MUG\\" is not in the product list, please provide a valid \\"productId\\"'"`;
